packages/gatsby/src/commands/build.js

```diff
@@ -137,14 +139,11 @@ module.exports = async function build(program: BuildArgs) {
     `BOOTSTRAP_QUERY_RUNNING_FINISHED`
   )
 
-  await db.saveState()
-
   await waitUntilAllJobsComplete()
 
-  // we need to save it again to make sure our latest state has been saved
-  await db.saveState()
-
-  const pagePaths = [...store.getState().pages.keys()]
+  const pagePaths = process.env.GATSBY_PAGE_BUILD_ON_DATA_CHANGES
+    ? await pageDataUtil.getChangedPageDataKeys(store.getState(), readState())
```

Comment:
We already read the state when we initialize, maybe we can save reference to required old state (webpack hash, pages and pageData). I think if we do at the beginning of build (before bootstrap), it should have old page data. I'm kinda wary about re-reading cache multiple times, who knows when it can get changed.

Reply:
Yep that makes sense 👍

Reply:
Ok that's done 🙇